File: backend/app/schemas/finanzas.py
"""
Schemas de Finanzas (Caja, Movimientos, Cuentas Bancarias).
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field


# ==================== CAJA ====================

class CajaBase(BaseModel):
    """Schema base de caja."""
    saldo_inicial: Decimal = Field(..., ge=0)
    observaciones_apertura: Optional[str] = None


class AbrirCajaRequest(CajaBase):
    """Request para abrir caja."""
    pass


class CerrarCajaRequest(BaseModel):
    """Request para cerrar caja."""
    saldo_real: Decimal = Field(..., ge=0)
    observaciones_cierre: Optional[str] = None


class CajaResponse(BaseModel):
    """Schema de respuesta de caja."""
    id: str
    numero: int
    fecha: date
    estado: str
    saldo_inicial: Decimal
    total_ingresos: Decimal
    total_egresos: Decimal
    saldo_final: Optional[Decimal] = None
    saldo_real: Optional[Decimal] = None
    diferencia: Optional[Decimal] = None
    abierta_por_id: str
    fecha_apertura: datetime
    cerrada_por_id: Optional[str] = None
    fecha_cierre: Optional[datetime] = None
    observaciones_apertura: Optional[str] = None
    observaciones_cierre: Optional[str] = None
    created_at: datetime

    # Calculados
    saldo_calculado: float
    abierta_por_nombre: Optional[str] = None
    cerrada_por_nombre: Optional[str] = None

    class Config:
        from_attributes = True


class CajaList(BaseModel):
    """Schema simplificado para listados."""
    id: str
    numero: int
    fecha: date
    estado: str
    saldo_inicial: Decimal
    total_ingresos: Decimal
    total_egresos: Decimal
    saldo_final: Optional[Decimal] = None
    diferencia: Optional[Decimal] = None

    class Config:
        from_attributes = True


# ==================== MOVIMIENTO CAJA ====================

class MovimientoCajaBase(BaseModel):
    """Schema base de movimiento de caja."""
    tipo: str  # ingreso, egreso
    categoria: str
    concepto: str = Field(..., min_length=1, max_length=255)
    descripcion: Optional[str] = None
    monto: Decimal = Field(..., gt=0)
    medio_pago: str = "efectivo"
    referencia: Optional[str] = None


class MovimientoCajaCreate(MovimientoCajaBase):
    """Request para crear movimiento de caja."""
    cliente_id: Optional[str] = None
    proveedor_id: Optional[str] = None
    pedido_id: Optional[str] = None


class MovimientoCajaResponse(MovimientoCajaBase):
    """Schema de respuesta de movimiento."""
    id: str
    caja_id: str
    cliente_id: Optional[str] = None
    proveedor_id: Optional[str] = None
    pedido_id: Optional[str] = None
    recibo_id: Optional[str] = None
    registrado_por_id: str
    anulado: bool
    fecha_anulacion: Optional[datetime] = None
    motivo_anulacion: Optional[str] = None
    created_at: datetime

    # Calculados
    cliente_nombre: Optional[str] = None
    proveedor_nombre: Optional[str] = None
    registrado_por_nombre: Optional[str] = None

    class Config:
        from_attributes = True


class AnularMovimientoRequest(BaseModel):
    """Request para anular movimiento."""
    motivo: str = Field(..., min_length=10)


# ==================== CUENTA BANCARIA ====================

class CuentaBancariaBase(BaseModel):
    """Schema base de cuenta bancaria."""
    nombre: str = Field(..., min_length=1, max_length=100)
    banco: str = Field(..., min_length=1, max_length=100)
    tipo_cuenta: str
    numero_cuenta: str = Field(..., min_length=1, max_length=50)
    cbu: Optional[str] = Field(None, min_length=22, max_length=22)
    alias: Optional[str] = None
    titular: str = Field(..., min_length=1, max_length=200)
    cuit_titular: Optional[str] = None
    notas: Optional[str] = None


class CuentaBancariaCreate(CuentaBancariaBase):
    """Request para crear cuenta bancaria."""
    saldo_actual: Decimal = Field(default=0, ge=0)
    es_principal: bool = False


class CuentaBancariaUpdate(BaseModel):
    """Request para actualizar cuenta bancaria."""
    nombre: Optional[str] = None
    alias: Optional[str] = None
    es_principal: Optional[bool] = None
    activa: Optional[bool] = None
    notas: Optional[str] = None


class CuentaBancariaResponse(CuentaBancariaBase):
    """Schema de respuesta de cuenta bancaria."""
    id: str
    saldo_actual: Decimal
    saldo_disponible: Optional[Decimal] = None
    activa: bool
    es_principal: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ==================== MOVIMIENTO BANCARIO ====================

class MovimientoBancarioCreate(BaseModel):
    """Request para crear movimiento bancario."""
    cuenta_id: str
    tipo: str
    concepto: str = Field(..., min_length=1, max_length=255)
    descripcion: Optional[str] = None
    monto: Decimal = Field(..., gt=0)
    fecha_movimiento: date
    fecha_valor: Optional[date] = None
    numero_comprobante: Optional[str] = None
    referencia_externa: Optional[str] = None
    cliente_id: Optional[str] = None
    proveedor_id: Optional[str] = None


class MovimientoBancarioResponse(BaseModel):
    """Schema de respuesta de movimiento bancario."""
    id: str
    cuenta_id: str
    tipo: str
    concepto: str
    descripcion: Optional[str] = None
    monto: Decimal
    saldo_anterior: Decimal
    saldo_posterior: Decimal
    fecha_movimiento: date
    fecha_valor: Optional[date] = None
    numero_comprobante: Optional[str] = None
    referencia_externa: Optional[str] = None
    cliente_id: Optional[str] = None
    proveedor_id: Optional[str] = None
    conciliado: bool
    fecha_conciliacion: Optional[datetime] = None
    registrado_por_id: str
    created_at: datetime

    # Calculados
    cuenta_nombre: Optional[str] = None
    cliente_nombre: Optional[str] = None
    proveedor_nombre: Optional[str] = None

    class Config:
        from_attributes = True


# ==================== RESUMEN FINANCIERO ====================

class ResumenCajaDiario(BaseModel):
    """Resumen diario de caja."""
    fecha: date
    caja_numero: Optional[int] = None
    estado: Optional[str] = None
    saldo_inicial: Decimal
    total_ingresos: Decimal
    total_egresos: Decimal
    saldo_actual: Decimal
    cantidad_movimientos: int


class ResumenFinanciero(BaseModel):
    """Resumen financiero general."""
    # Caja
    caja_actual: Optional[ResumenCajaDiario] = None

    # Totales del período
    total_ingresos_periodo: Decimal
    total_egresos_periodo: Decimal
    balance_periodo: Decimal

    # Por categoría
    ingresos_por_categoria: dict
    egresos_por_categoria: dict

    # Cuentas bancarias
    total_en_bancos: Decimal


# ==================== CONSTANTES ====================

CATEGORIAS_INGRESO = [
    {"value": "venta", "label": "Venta"},
    {"value": "cobro_cliente", "label": "Cobro a Cliente"},
    {"value": "otro_ingreso", "label": "Otro Ingreso"},
]

CATEGORIAS_EGRESO = [
    {"value": "pago_proveedor", "label": "Pago a Proveedor"},
    {"value": "pago_empleado", "label": "Pago a Empleado"},
    {"value": "gasto_operativo", "label": "Gasto Operativo"},
    {"value": "compra_insumos", "label": "Compra de Insumos"},
    {"value": "servicio", "label": "Servicio (Luz, Agua, Gas)"},
    {"value": "impuesto", "label": "Impuesto"},
    {"value": "retiro", "label": "Retiro"},
    {"value": "otro_egreso", "label": "Otro Egreso"},
]

TIPOS_CUENTA_BANCARIA = [
    {"value": "caja_ahorro", "label": "Caja de Ahorro"},
    {"value": "cuenta_corriente", "label": "Cuenta Corriente"},
]

TIPOS_MOVIMIENTO_BANCO = [
    {"value": "deposito", "label": "Depósito"},
    {"value": "extraccion", "label": "Extracción"},
    {"value": "transferencia_entrada", "label": "Transferencia Recibida"},
    {"value": "transferencia_salida", "label": "Transferencia Enviada"},
    {"value": "debito_automatico", "label": "Débito Automático"},
    {"value": "credito", "label": "Crédito"},
    {"value": "comision", "label": "Comisión Bancaria"},
    {"value": "cheque_emitido", "label": "Cheque Emitido"},
    {"value": "cheque_depositado", "label": "Cheque Depositado"},
]

File: frontend/src/types/finanzas.ts
/**
 * Tipos para el módulo de Finanzas
 */

// Estados y enums
export type EstadoCaja = 'abierta' | 'cerrada';
export type TipoMovimientoCaja = 'ingreso' | 'egreso';
export type CategoriaMovimiento =
  | 'venta'
  | 'cobro_cliente'
  | 'otro_ingreso'
  | 'pago_proveedor'
  | 'pago_empleado'
  | 'gasto_operativo'
  | 'compra_insumos'
  | 'servicio'
  | 'impuesto'
  | 'retiro'
  | 'otro_egreso';

export type TipoCuentaBancaria = 'caja_ahorro' | 'cuenta_corriente';
export type TipoMovimientoBanco =
  | 'deposito'
  | 'extraccion'
  | 'transferencia_entrada'
  | 'transferencia_salida'
  | 'debito_automatico'
  | 'credito'
  | 'comision'
  | 'interes'
  | 'cheque_emitido'
  | 'cheque_depositado';

// Caja
export interface Caja {
  id: string;
  numero: number;
  fecha: string;
  estado: EstadoCaja;
  saldo_inicial: number;
  total_ingresos: number;
  total_egresos: number;
  saldo_final: number | null;
  saldo_real: number | null;
  diferencia: number | null;
  abierta_por_id: string;
  fecha_apertura: string;
  cerrada_por_id: string | null;
  fecha_cierre: string | null;
  observaciones_apertura: string | null;
  observaciones_cierre: string | null;
  created_at: string;
  saldo_calculado: number;
  abierta_por_nombre: string | null;
  cerrada_por_nombre: string | null;
}

export interface CajaList {
  id: string;
  numero: number;
  fecha: string;
  estado: EstadoCaja;
  saldo_inicial: number;
  total_ingresos: number;
  total_egresos: number;
  saldo_final: number | null;
  diferencia: number | null;
}

export interface AbrirCajaRequest {
  saldo_inicial: number;
  observaciones_apertura?: string | null;
}

export interface CerrarCajaRequest {
  saldo_real: number;
  observaciones_cierre?: string | null;
}

// Movimiento de Caja
export interface MovimientoCaja {
  id: string;
  caja_id: string;
  tipo: TipoMovimientoCaja;
  categoria: CategoriaMovimiento;
  concepto: string;
  descripcion: string | null;
  monto: number;
  medio_pago: string;
  referencia: string | null;
  cliente_id: string | null;
  proveedor_id: string | null;
  pedido_id: string | null;
  recibo_id: string | null;
  registrado_por_id: string;
  anulado: boolean;
  fecha_anulacion: string | null;
  motivo_anulacion: string | null;
  created_at: string;
  cliente_nombre: string | null;
  proveedor_nombre: string | null;
  registrado_por_nombre: string | null;
}

export interface MovimientoCajaCreate {
  tipo: TipoMovimientoCaja;
  categoria: CategoriaMovimiento;
  concepto: string;
  descripcion?: string | null;
  monto: number;
  medio_pago?: string;
  referencia?: string | null;
  cliente_id?: string | null;
  proveedor_id?: string | null;
  pedido_id?: string | null;
}

// Cuenta Bancaria
export interface CuentaBancaria {
  id: string;
  nombre?: string;
  banco: string;
  tipo_cuenta: TipoCuentaBancaria | string;
  numero_cuenta: string;
  cbu: string | null;
  alias: string | null;
  titular: string;
  cuit_titular?: string | null;
  saldo_actual: number;
  saldo_inicial?: number;
  saldo_disponible?: number | null;
  activa?: boolean;
  activo?: boolean;
  es_principal?: boolean;
  notas?: string | null;
  moneda?: string;
  created_at?: string;
  updated_at?: string | null;
}

// Movimiento Bancario
export interface MovimientoBancario {
  id: string;
  cuenta_id: string;
  tipo: TipoMovimientoBanco | string;
  concepto?: string;
  descripcion?: string | null;
  monto: number;
  fecha: string;
  saldo_anterior?: number;
  saldo_posterior?: number;
  fecha_movimiento?: string;
  fecha_valor?: string | null;
  numero_comprobante?: string | null;
  referencia_externa?: string | null;
  cliente_id?: string | null;
  proveedor_id?: string | null;
  conciliado?: boolean;
  fecha_conciliacion?: string | null;
  registrado_por_id?: string;
  categoria?: string;
  created_at?: string;
  cuenta_nombre?: string | null;
  cliente_nombre?: string | null;
  proveedor_nombre?: string | null;
}

// Resumen Financiero
export interface ResumenCajaDiario {
  fecha: string;
  caja_numero: number | null;
  estado: string | null;
  saldo_inicial: number;
  total_ingresos: number;
  total_egresos: number;
  saldo_actual: number;
  cantidad_movimientos: number;
}

export interface ResumenCaja {
  ingresos: number;
  egresos: number;
  saldo_actual: number;
  total_movimientos: number;
  movimientos_por_categoria?: Record<string, number>;
}

export interface ResumenBancos {
  ingresos: number;
  egresos: number;
  saldo_total: number;
  total_movimientos: number;
}

export interface MovimientoResumen {
  fecha: string;
  tipo: string;
  monto: number;
  descripcion?: string;
  categoria?: string;
  origen: 'caja' | 'banco';
}

export interface ResumenFinanciero {
  caja_actual?: ResumenCajaDiario | null;
  total_ingresos_periodo?: number;
  total_egresos_periodo?: number;
  balance_periodo?: number;
  ingresos_por_categoria?: Record<string, number>;
  egresos_por_categoria?: Record<string, number>;
  total_en_bancos?: number;
  // Nueva estructura alternativa
  caja?: ResumenCaja;
  bancos?: ResumenBancos;
  ultimos_movimientos?: MovimientoResumen[];
}

// Constantes
export const CATEGORIAS_INGRESO = [
  { value: 'venta', label: 'Venta' },
  { value: 'cobro_cliente', label: 'Cobro a Cliente' },
  { value: 'otro_ingreso', label: 'Otro Ingreso' },
];

export const CATEGORIAS_EGRESO = [
  { value: 'pago_proveedor', label: 'Pago a Proveedor' },
  { value: 'pago_empleado', label: 'Pago a Empleado' },
  { value: 'gasto_operativo', label: 'Gasto Operativo' },
  { value: 'compra_insumos', label: 'Compra de Insumos' },
  { value: 'servicio', label: 'Servicio (Luz, Agua, Gas)' },
  { value: 'impuesto', label: 'Impuesto' },
  { value: 'retiro', label: 'Retiro' },
  { value: 'otro_egreso', label: 'Otro Egreso' },
];

export const MEDIOS_PAGO_CAJA = [
  { value: 'efectivo', label: 'Efectivo' },
  { value: 'transferencia', label: 'Transferencia' },
  { value: 'tarjeta_debito', label: 'Tarjeta Débito' },
  { value: 'tarjeta_credito', label: 'Tarjeta Crédito' },
  { value: 'cheque', label: 'Cheque' },
  { value: 'mercado_pago', label: 'Mercado Pago' },
];

export const TIPOS_CUENTA_BANCARIA = [
  { value: 'caja_ahorro', label: 'Caja de Ahorro' },
  { value: 'cuenta_corriente', label: 'Cuenta Corriente' },
];
